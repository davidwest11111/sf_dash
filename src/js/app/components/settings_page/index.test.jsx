import React from 'react';
import { shallow } from 'enzyme';
import shallowToJson from 'enzyme-to-json';

import LocalStorage, { WEATHER_UNITS_KEY, WEATHER_KEY_KEY } from 'app/utils/local_storage';
import DarkSky from 'app/integrations/darksky';

import Settings from './index';


describe('outputs the expected tree when', () => {
  let wrapper;

  test('(default)', () => {
    wrapper = shallow((
      <Settings />
    ));
  });

  afterEach(() => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});

describe('instance methods', () => {
  let localStorageGetSpy;
  let localStorageSetSpy;
  let dumpLocalStorageSpy;
  let wrapper;
  let instance;

  beforeEach(() => {
    localStorageGetSpy = jest.spyOn(LocalStorage, 'get')
      .mockReturnValue('foo');
    localStorageSetSpy = jest.spyOn(LocalStorage, 'set').mockReturnValue();
    dumpLocalStorageSpy = jest.spyOn(LocalStorage, 'dump').mockReturnValue();
    wrapper = shallow((
      <Settings />
    ));
    instance = wrapper.instance();
  });

  test('getInitialState', () => {
    expect(instance.getInitialState()).toEqual({
      weatherKeyIsValid: undefined,
      weatherKeyValue: 'foo',
      weatherUnits: 'foo',
    });
    expect(localStorageGetSpy).toHaveBeenCalledWith(WEATHER_KEY_KEY);
  });

  test('onWeatherKeyChange', () => {
    const value = 'abc';

    instance.onWeatherKeyChange({ target: { value } });

    expect(instance.getInitialState()).toEqual({
      weatherKeyIsValid: undefined,
      weatherKeyValue: 'foo',
      weatherUnits: 'foo',
    });
    expect(wrapper.state().weatherKeyValue).toEqual(value);
  });

  describe('validateAndSetKey', () => {
    let weatherKeyValue;

    beforeEach(() => {
      weatherKeyValue = 'abc';
      wrapper.setState({ weatherKeyValue });
    });

    test('is valid key', async () => {
      const spy = jest.spyOn(DarkSky, 'isValidKey').mockReturnValue(true);

      await instance.validateAndSetKey({ preventDefault: () => {} });

      expect(spy).toHaveBeenCalledWith(weatherKeyValue);
      expect(wrapper.state().weatherKeyIsValid).toEqual(true);
      expect(localStorageSetSpy)
        .toHaveBeenCalledWith(WEATHER_KEY_KEY, weatherKeyValue);
    });

    test('is invalid key', async () => {
      const spy = jest.spyOn(DarkSky, 'isValidKey').mockReturnValue(false);

      await instance.validateAndSetKey({ preventDefault: () => {} });

      expect(spy).toHaveBeenCalledWith(weatherKeyValue);
      expect(wrapper.state().weatherKeyIsValid).toEqual(false);
    });
  });

  test('setWeatherUnits', () => {
    const units = 'c';

    instance.setWeatherUnits(units);

    expect(wrapper.state().weatherUnits).toEqual(units);
    expect(localStorageSetSpy).toHaveBeenCalledWith(WEATHER_UNITS_KEY, units);
  });

  test('dumpLocalStorageAndResetState', () => {
    instance.dumpLocalStorageAndResetState();

    expect(wrapper.state()).toEqual(instance.getInitialState());
    expect(dumpLocalStorageSpy).toHaveBeenCalledWith();
  });


  test('button clicks call setWeatherUnits', () => {
    const spy = jest.spyOn(instance, 'setWeatherUnits');

    wrapper.find('Button').at(2).simulate('click');
    expect(spy).toHaveBeenCalledWith('');

    wrapper.find('Button').at(3).simulate('click');
    expect(spy).toHaveBeenCalledWith('f');

    wrapper.find('Button').at(4).simulate('click');
    expect(spy).toHaveBeenCalledWith('c');
  });
});
