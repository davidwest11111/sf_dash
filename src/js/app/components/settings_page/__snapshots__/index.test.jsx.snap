// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`outputs the expected tree when (default) 1`] = `
<div
  className=""
>
  <h1>
    Settings
  </h1>
  <Row
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
      xs="6"
    >
      <Card
        className="mb-4"
        tag="div"
      >
        <CardHeader
          tag="div"
        >
          <h3
            className="m-0"
          >
            Global
          </h3>
        </CardHeader>
        <CardBody
          tag="div"
        >
          <DarkLightModeSelector
            className=""
          />
          <hr />
          <p>
            <Button
              color="danger"
              onClick={[Function]}
              tag="button"
            >
              Dump All Local Storage
            </Button>
          </p>
          <p
            className="small"
          >
            This will remove stored bus stops, your DarkSky key, and the weather units.
          </p>
        </CardBody>
      </Card>
    </Col>
  </Row>
  <Row
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
      xs="6"
    >
      <Card
        className="mb-4"
        tag="div"
      >
        <CardHeader
          tag="div"
        >
          <h3
            className="m-0"
          >
            Bus
          </h3>
        </CardHeader>
        <CardBody
          tag="div"
        >
          <DisplayLocalStorageData
            className=""
            localStorageKey="routeStopTags"
          />
          <hr />
          <SetSampleLocalStorageData
            className=""
            localStorageKey="routeStopTags"
          />
        </CardBody>
      </Card>
    </Col>
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
      xs="6"
    >
      <Card
        className="mb-4"
        tag="div"
      >
        <CardHeader
          tag="div"
        >
          <h3
            className="m-0"
          >
            Weather
          </h3>
        </CardHeader>
        <CardBody
          tag="div"
        >
          <Form
            onSubmit={[Function]}
            tag="form"
          >
            <FormGroup
              tag="div"
            >
              <Label
                for="weather-key"
                tag="label"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                DarkSky API Key:
              </Label>
              <Input
                className=""
                id="weather-key"
                onChange={[Function]}
                type="text"
                value=""
              />
            </FormGroup>
            <Button
              color="primary"
              tag="button"
              type="submit"
            >
              Validate & Store Key
            </Button>
          </Form>
          <hr />
          <FormGroup
            tag="div"
          >
            <Label
              className="mr-2"
              tag="label"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              Weather Units:
            </Label>
            <ButtonGroup
              role="group"
              tag="div"
            >
              <Button
                color="primary"
                onClick={[Function]}
                tag="button"
              >
                Auto
              </Button>
              <Button
                color="secondary"
                onClick={[Function]}
                tag="button"
              >
                F
              </Button>
              <Button
                color="secondary"
                onClick={[Function]}
                tag="button"
              >
                C
              </Button>
            </ButtonGroup>
          </FormGroup>
        </CardBody>
      </Card>
    </Col>
  </Row>
</div>
`;
